src/model.jl

```diff
@@ -66,21 +66,68 @@ function _make_bus_demand_flexibility_amount(
+    if demand_flexibility.enable_doe_flexibility
+        doe_flex_pct = Matrix(
+            demand_flexibility.doe_flex_amt[start_index:end_index, 2:end]
+        )
+        doe_flex_mw = doe_flex_pct .*= permutedims(bus_demand[sets.doe_flexible_bus_idx, :])
+        for i in 1:length(sets.doe_flexible_bus_idx)
+            simulation_demand_flex_amt_up[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[:, i]
+            simulation_demand_flex_amt_dn[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[:, i]
+        end
+    end 
+    
+    # replace columns with user input numbers
+    csv_flex_amt_up = Matrix(
         demand_flexibility.flex_amt_up[start_index:end_index, 2:end]
     )
-    # Demand flexibility down profile
-    simulation_demand_flex_amt_dn = Matrix(
+    csv_flex_amt_dn = Matrix(
         demand_flexibility.flex_amt_dn[start_index:end_index, 2:end]
     )
+    
+    csv_flex_cost_up = Matrix(
+        demand_flexibility.cost_up[:, 2:end]
+    )
+    csv_flex_cost_dn = Matrix(
+        demand_flexibility.cost_dn[:, 2:end]
+    )
 
-    # percentage flexibility from input
-    if demand_flexibility.enable_doe_flexibility
-        simulation_demand_flex_amt_up .*= permutedims(bus_demand[sets.flexible_bus_idx, :])
-        simulation_demand_flex_amt_dn .*= permutedims(bus_demand[sets.flexible_bus_idx, :])
+    for i in 1:length(sets.csv_flexible_bus_idx)
+        simulation_demand_flex_amt_up[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_up[:, i]
+        simulation_demand_flex_amt_dn[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_dn[:, i]
+        
+        new_demand_flex_cost_up[:, sets.csv_flexible_bus_idx[i]] = csv_flex_cost_up[:, i]
+        new_demand_flex_cost_dn[:, sets.csv_flexible_bus_idx[i]] = csv_flex_cost_dn[:, i]
     end
 
+    # remove non-flexible columns
+    simulation_demand_flex_amt_up = simulation_demand_flex_amt_up[:, sets.flexible_bus_idx]
+    simulation_demand_flex_amt_dn = simulation_demand_flex_amt_dn[:, sets.flexible_bus_idx]
+    new_demand_flex_cost_up = new_demand_flex_cost_up[:, sets.flexible_bus_idx]
+    new_demand_flex_cost_dn = new_demand_flex_cost_dn[:, sets.flexible_bus_idx]
+    
+    # construct new cost Dataframe with appropriate dimension
+    demand_flexibility.cost_up = DataFrames.DataFrame(new_demand_flex_cost_up, :auto)
+    demand_flexibility.cost_dn = DataFrames.DataFrame(new_demand_flex_cost_dn, :auto)
+    
+    DataFrames.rename!(demand_flexibility.cost_dn, Symbol.(sets.flexible_bus_idx))
+    DataFrames.rename!(demand_flexibility.cost_dn, Symbol.(sets.flexible_bus_idx))
+    DataFrames.insertcols!(
+        demand_flexibility.cost_up, 1, :"UTC Time" => demand_flexibility.flex_amt_up[!, "UTC Time"]
+    )
+    DataFrames.insertcols!(
+        demand_flexibility.cost_dn, 1, :"UTC Time" => demand_flexibility.flex_amt_up[!, "UTC Time"]
+    )    
+    
     return (
         permutedims(simulation_demand_flex_amt_up),
         permutedims(simulation_demand_flex_amt_dn),
```

Comment:
If we don't end up making the changes I suggested for this function, I believe we'll still need to return the updated `demand_flexibility`.

Reply:
I am going to fix the mutable problem. but in julia do I need to also return an updated struct/DF after changing it in a function to keep the change?

Reply:
I believe so? I haven't tested it (though we could probably do a pretty simple test), but I believe the function creates a copy of the `DemandFlexibility` object that is local to the function. If we wanted to keep changes made to the variable, I believe we need to return it from the function

Reply:
hmm that's annoying if so. even the ancient C doesn't do this.

Reply:
Ah, it appears that you're right: 

http://web.mit.edu/julia_v0.6.2/julia/share/doc/julia/html/en/manual/functions.html#:~:text=Julia%20function%20arguments%20follow%20a,identical%20to%20the%20passed%20values.

Though if the changes are made, this doesn't seem like it will come up.